test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/MultiTestRunFinalization/MultiTestRunFinalizationManagerTests.cs

```diff
@@ -0,0 +1,625 @@
+            {
+                try
+                {
+                    for (int i = 0; i < 100; ++i)
+                    {
+                        Task.Delay(100).Wait();
+                        Console.WriteLine($"Iteration: {i}");
+                        logger.SendMessage(TestMessageLevel.Informational, $"Iteration: {i}");
+
+                        cancellation.ThrowIfCancellationRequested();
+                        progress.Report(i + 1);
+
+                        if (i == 3)
+                        {
+                            cancellationTokenSource.Cancel();
+                            Task.Delay(500).Wait();
+                        }
+                    }
+                }
+                finally
+                {
+                    innerTaskCompletionSource.TrySetResult(null);
+                }
+
+                return Task.FromResult(outputAttachments);
+            });
+
+            // act
+            await manager.FinalizeMultiTestRunAsync(mockRequestData.Object, inputAttachments, mockEventsHandler.Object, cancellationTokenSource.Token);
+            Console.WriteLine("Finalization done");
+            await innerTaskCompletionSource.Task;
+
+            // assert
+            VerifyCompleteEvent(true, false, inputAttachments[0]);
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.IsAny<MultiTestRunFinalizationProgressEventArgs>()), Times.Exactly(4));
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.Is<MultiTestRunFinalizationProgressEventArgs>(a => VerifyProgressArgs(a, 1))));
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.Is<MultiTestRunFinalizationProgressEventArgs>(a => VerifyProgressArgs(a, 2))));
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.Is<MultiTestRunFinalizationProgressEventArgs>(a => VerifyProgressArgs(a, 3))));
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.Is<MultiTestRunFinalizationProgressEventArgs>(a => VerifyProgressArgs(a, 4))));
+            mockEventsHandler.Verify(h => h.HandleLogMessage(TestMessageLevel.Informational, "Finalization was cancelled."));
+            mockAttachmentHandler1.Verify(h => h.GetExtensionUris());
+            mockAttachmentHandler2.Verify(h => h.GetExtensionUris(), Times.Never);
+            mockAttachmentHandler1.Verify(h => h.ProcessAttachmentSetsAsync(It.Is<ICollection<AttachmentSet>>(c => c.Count == 1 && c.Contains(inputAttachments[0])), It.IsAny<IProgress<int>>(), It.IsAny<IMessageLogger>(), cancellationTokenSource.Token));
+            mockAttachmentHandler2.Verify(h => h.ProcessAttachmentSetsAsync(It.IsAny<ICollection<AttachmentSet>>(), It.IsAny<IProgress<int>>(), It.IsAny<IMessageLogger>(), It.IsAny<CancellationToken>()), Times.Never);
+
+            VerifyMetrics(1, 1, "Canceled");
```

Comment:
>1 [](start = 26, length = 1)

consider using named parameters for all Verify calls.

Reply:
fixed